pkg/placement/placement.go

```diff
@@ -79,18 +74,37 @@ func (p *Service) ReportDaprStatus(srv placementv1pb.Placement_ReportDaprStatusS
 
 		req, err := srv.Recv()
 		if err != nil {
-			p.hostsLock.Lock()
-			p.RemoveHost(srv)
-			p.ProcessRemovedHost(id)
-			log.Infof("host removed: %s", id)
-			p.hostsLock.Unlock()
+			func() {
+				p.hostsLock.Lock()
+				p.RemoveHost(srv)
+				p.ProcessRemovedHost(id)
+				log.Infof("host removed: %s", id)
+				p.hostsLock.Unlock()
+			}()
 			continue
 		}
 
 		p.ProcessHost(req)
 	}
 }
 
+func (p *Service) getIdFromContext(ctx context.Context, srv placementv1pb.Placement_ReportDaprStatusServer) (string, error) {
+	p.hostsLock.Lock()
+	defer p.hostsLock.Unlock()
+
+	md, _ := metadata.FromIncomingContext(ctx)
+	v := md.Get("id")
+	if len(v) == 0 {
+		return "", errors.New("id header not found in metadata")
+	}
+
+	id := v[0]
+	p.hosts = append(p.hosts, srv)
+	log.Infof("host added: %s", id)
```

Comment:
The method is a get ID method, but here we are adding the host.

My suggestion is to either change the method name to something like `addHost` or take the hosts add logic outside of this method.

Reply:
OK,since the opeation of adding host should be protected by the host lock, it should not take out of this method.

So I renamed this method from getIdFromContext to addHost.